README.md
Iris2
=====

[![Build Status](https://travis-ci.org/go-iris2/iris2.svg?branch=master)](https://travis-ci.org/go-iris2/iris2)
[![Go Report Card](https://goreportcard.com/badge/github.com/go-iris2/iris2)](https://goreportcard.com/report/github.com/go-iris2/iris2)
[![GoDoc](https://godoc.org/github.com/go-iris2/iris2?status.svg)](https://godoc.org/github.com/go-iris2/iris2)
[![codecov](https://codecov.io/gh/go-iris2/iris2/branch/master/graph/badge.svg)](https://codecov.io/gh/go-iris2/iris2)
[![codebeat badge](https://codebeat.co/badges/75a18e68-c2da-4ec0-a15d-335ab6836efc)](https://codebeat.co/projects/github-com-go-iris2-iris2)

Iris2 is a fork of the original [Iris](https://github.com/kataras/iris) framework. As Iris, Iris2 is an efficient and complete toolbox with robust set of features.<br/>Write <b>your own</b>
<b>perfect high-performance web applications</b> <br/>with unlimited potentials and <b>portability</b>.<br/>
Iris2 aims for a *stable* API and be easy to use.

Installation
------------

The only requirement is the [Go Programming Language](https://golang.org/dl/), version 1.8+

```bash
$ go get github.com/go-iris2/iris2
```


Overview
--------

```go
package main

import (
	"net/http"

	"github.com/go-iris2/iris2"
	"github.com/go-iris2/iris2/adaptors/view"
)

func main() {
	app := iris2.New()
	app.Adapt(iris2.Devlogger())

	// 5 template engines are supported out-of-the-box:
	//
	// - standard html/template
	// - amber
	// - django
	// - handlebars
	// - pug(jade)
	//
	// Use the html standard engine for all files inside "./views" folder with extension ".html"
	templates := view.HTML("./views", ".html")
	app.Adapt(templates)

	// http://localhost:6200
	// Method: "GET"
	// Render ./views/index.html
	app.Get("/", func(ctx *iris2.Context) {
		ctx.Render("index.html", nil)
	})

	// Group routes, optionally: share middleware, template layout and custom http errors.
	userAPI := app.Party("/users", userAPIMiddleware).
		Layout("layouts/userLayout.html")
	{
		// Fire userNotFoundHandler when Not Found
		// inside http://localhost:6200/users/*anything
		userAPI.OnError(404, userNotFoundHandler)

		// http://localhost:6200/users
		// Method: "GET"
		userAPI.Get("/", getAllHandler)

		// http://localhost:6200/users/42
		// Method: "GET"
		userAPI.Get("/:id", getByIDHandler)

		// http://localhost:6200/users
		// Method: "POST"
		userAPI.Post("/", saveUserHandler)
	}

	// Start the server at 127.0.0.1:6200
	app.Listen(":6200")
}

func getByIDHandler(ctx *iris2.Context) {
	// take the :id from the path, parse to integer
	// and set it to the new userID local variable.
	userID, _ := ctx.ParamInt("id")

	// userRepo, imaginary database service <- your only job.
	user := userRepo.GetByID(userID)

	// send back a response to the client,
	// .JSON: content type as application/json; charset="utf-8"
	// http.StatusOK: with 200 http status code.
	//
	// send user as it is or make use of any json valid golang type,
	// like the iris2.Map{"username" : user.Username}.
	ctx.JSON(http.StatusOK, user)
}
```

Versioning
----------

Iris2 adheres to [Semantic Versioning](http://semver.org/).

See the [Changelog](CHANGELOG.md).

License
-------

Unless otherwise noted, the source files are distributed
under the MIT License found in the [LICENSE](LICENSE).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

**Upgrading**: Use: `go get -u github.com/go-iris2/iris2` and you are done.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
### Added
- File storage for sessions
- Leveldb storage for sessions

### Changed
- Fork from kataras/iris to go-iris2/iris2 and rename (`4b71e60`)
- Default routing uses httprouter

### Removed
- Option for gorillamux in favor of user-friendlyness

### Fixed
- Sessions bound to IP-addresses
- Possible session-collisions


doc.go
// Copyright (c) 2016-2017 Gerasimos Maropoulos
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

/*
Package iris2 provides efficient and well-designed tools with robust set of features to
create your own perfect high performance web application
with unlimited potentials and portability.

Source code and other details for the project are available at GitHub:

   https://github.com/go-iris2/iris2

Note: This package is under active development status.
Each month a new version is releasing
to adapt the latest web trends and technologies.

Basic HTTP API

Iris is a very pluggable ecosystem,
router can be customized by adapting a 'RouterBuilderPolicy && RouterReversionPolicy'.

All HTTP methods are supported, users can register handlers for same paths on different methods.
The first parameter is the HTTP Method,
second parameter is the request path of the route,
third variadic parameter should contains one or more iris2.Handler/HandlerFunc executed
by the registered order when a user requests for that specific resouce path from the server.

Example code:


        app := iris2.New()

        app.Handle("GET", "/about", aboutHandler)

        type aboutHandler struct {}
        func (a aboutHandler) Serve(ctx *iris2.Context){
          ctx.HTML("Hello from /about, executed from an iris2.Handler")
        }

        app.HandleFunc("GET", "/contact", func(ctx *iris2.Context){
          ctx.HTML(http.StatusOK, "Hello from /contact, executed from an iris2.HandlerFunc")
        })


In order to make things easier for the user, Iris provides functions for all HTTP Methods.
The first parameter is the request path of the route,
second variadic parameter should contains one or more iris2.HandlerFunc executed
by the registered order when a user requests for that specific resouce path from the server.

Example code:


      app := iris2.New()

      // Method: "GET"
      app.Get("/", handler)

      // Method: "POST"
      app.Post("/", handler)

      // Method: "PUT"
      app.Put("/", handler)

      // Method: "DELETE"
      app.Delete("/", handler)

      // Method: "OPTIONS"
      app.Options("/", handler)

      // Method: "TRACE"
      app.Trace("/", handler)

      // Method: "CONNECT"
      app.Connect("/", handler)

      // Method: "HEAD"
      app.Head("/", handler)

      // Method: "PATCH"
      app.Patch("/", handler)

      // register the route for all HTTP Methods
      app.Any("/", handler)

      func handler(ctx *iris2.Context){
        ctx.Writef("Hello from method: %s and path: %s", ctx.Method(), ctx.Path())
      }


Parameterized Path


Path Parameters' syntax depends on the selected router.
This is the only difference between the routers, the registered path form, the API remains the same for both.

      package main

      import (
        "github.com/go-iris2/iris2"
      )

      func main() {
        app := iris2.New()

        app.OnError(http.StatusNotFound, func(ctx *iris2.Context){
          ctx.HTML(http.StatusNotFound, "<h1> custom http error page </h1>")
        })


        app.Get("/healthcheck", h)

        gamesMiddleware := func(ctx *iris2.Context) {
          println(ctx.Method() + ": " + ctx.Path())
          ctx.Next()
        }

        games:= app.Party("/games", gamesMiddleware)
        { // braces are optional of course, it's just a style of code
        	 games.Get("/:gameID/clans", h)
        	 games.Get("/:gameID/clans/clan/:publicID", h)
        	 games.Get("/:gameID/clans/search", h)

        	 games.Put("/:gameID/players/:publicID", h)
        	 games.Put("/:gameID/clans/clan/:publicID", h)

        	 games.Post("/:gameID/clans", h)
        	 games.Post("/:gameID/players", h)
        	 games.Post("/:gameID/clans/:publicID/leave", h)
        	 games.Post("/:gameID/clans/:clanPublicID/memberships/application", h)
        	 games.Post("/:gameID/clans/:clanPublicID/memberships/application/:action", h)
        	 games.Post("/:gameID/clans/:clanPublicID/memberships/invitation", h)
        	 games.Post("/:gameID/clans/:clanPublicID/memberships/invitation/:action", h)
        	 games.Post("/:gameID/clans/:clanPublicID/memberships/delete", h)
        	 games.Post("/:gameID/clans/:clanPublicID/memberships/promote", h)
        	 games.Post("/:gameID/clans/:clanPublicID/memberships/demote", h)
        }

        app.Get("/anything/*anythingparameter", func(ctx *iris2.Context){
          s := ctx.Param("anythingparameter")
          ctx.Writef("The path after /anything is: %s",s)
        })

        mysubdomain:= app.Party("mysubdomain.")
        // http://mysubdomain.myhost.com/
        mysudomain.Get("/", h)

        app.Listen("myhost.com:80")
      }

      func h(ctx *iris2.Context) {
      	ctx.HTML(http.StatusOK, "<h1>Path<h1/>"+ctx.Path())
      }


Grouping Routes


A set of routes that are being groupped by path prefix can (optionally) share the same middleware handlers and template layout.
A group can have a nested group too.

`.Party` is being used to group routes, developers can declare an unlimited number of (nested) groups.


Example code:


      users:= app.Party("/users", myAuthHandler)

      // http://myhost.com/users/42/profile
      users.Get("/:userid/profile", userProfileHandler)
      // http://myhost.com/users/messages/1
      users.Get("/inbox/:messageid", userMessageHandler)

      app.Listen("myhost.com:80")



Custom HTTP Errors


With iris users are able to register their own handlers for http statuses like 404 not found, 500 internal server error and so on.

Example code:

      // when 404 then render the template $templatedir/errors/404.html
      // *read below for information about the view engine.*
      app.OnError(http.StatusNotFound, func(ctx *iris2.Context){
        ctx.RenderWithstatus(http.StatusNotFound, "errors/404.html", nil)
      })

      app.OnError(500, func(ctx *iris2.Context){
        // ...
      })


Custom http errors can be also be registered to a specific group of routes.

Static Files

      // Favicon serves static favicon
      // accepts 2 parameters, second is optional
      // favPath (string), declare the system directory path of the __.ico
      // requestPath (string), it's the route's path, by default this is the "/favicon.ico" because some browsers tries to get this by default first,
      // you can declare your own path if you have more than one favicon (desktop, mobile and so on)
      //
      // this func will add a route for you which will static serve the /yuorpath/yourfile.ico to the /yourfile.ico (nothing special that you can't handle by yourself)
      // Note that you have to call it on every favicon you have to serve automatically (desktop, mobile and so on)
      //
      // panics on error
      Favicon(favPath string, requestPath ...string) RouteInfo

      // StaticContent serves bytes, memory cached, on the reqPath
      // a good example of this is how the websocket server uses that to auto-register the /iris-ws.js
      StaticContent(reqPath string, cType string, content []byte) RouteInfo

      // StaticHandler returns a new Handler which serves static files
      StaticHandler(reqPath string, systemPath string, showList bool, enableGzip bool, exceptRoutes ...RouteInfo) HandlerFunc

      // StaticWeb returns a handler that serves HTTP requests
      // with the contents of the file system rooted at directory.
      //
      // first parameter: the route path
      // second parameter: the system directory
      // third OPTIONAL parameter: the exception routes
      //      (= give priority to these routes instead of the static handler)
      // for more options look iris2.StaticHandler.
      //
      //     iris2.StaticWeb("/static", "./static")
      //
      // As a special case, the returned file server redirects any request
      // ending in "/index.html" to the same path, without the final
      // "index.html".
      //
      // StaticWeb calls the StaticHandler(reqPath, systemPath, listingDirectories: false, gzip: false ).
      StaticWeb(reqPath string, systemPath string, exceptRoutes ...RouteInfo) RouteInfo

      // StaticEmbedded  used when files are distributed inside the app executable, using go-bindata mostly
      // First parameter is the request path, the path which the files in the vdir will be served to, for example "/static"
      // Second parameter is the (virtual) directory path, for example "./assets"
      // Third parameter is the Asset function
      // Forth parameter is the AssetNames function
      StaticEmbedded(reqPath string, vdir string, assetFn func(name string) ([]byte, error), namesFn func() []string) RouteInfo

Example code:


      package main

      import (
      	"github.com/go-iris2/iris2"
      )

      func main() {

      	app := iris2.New()

      	app.Favicon("./static/favicons/iris_favicon_32_32.ico")
      	// This will serve the ./static/favicons/iris_favicon_32_32.ico to: 127.0.0.1:8080/favicon.ico

      	// app.Favicon("./static/favicons/iris_favicon_32_32.ico", "/favicon_32_32.ico")
      	// This will serve the ./static/favicons/iris_favicon_32_32.ico to: 127.0.0.1:8080/favicon_32_32.ico

      	app.Get("/", func(ctx *iris2.Context) {
      		ctx.HTML(http.StatusOK, "You should see the favicon now at the side of your browser.")
      	})

      	app.Listen(":8080")
      }



Middleware Ecosystem

Middleware is just a concept of ordered chain of handlers.
Middleware can be registered globally, per-party, per-subdomain and per-route.


Example code:

      // globally
      // before any routes, appends the middleware to all routes
      app.UseFunc(func(ctx *iris2.Context){
         // ... any code here

         ctx.Next() // in order to continue to the next handler,
         // if that is missing then the next in chain handlers will be not executed,
         // useful for authentication middleware
      })

      // globally
      // after or before any routes, prepends the middleware to all routes
      app.UseGlobalFunc(handlerFunc1, handlerFunc2, handlerFunc3)

      // per-route
      app.Post("/login", authenticationHandler, loginPageHandler)

      // per-party(group of routes)
      users := app.Party("/users", usersMiddleware)
      users.Get("/", usersIndex)

      // per-subdomain
      mysubdomain := app.Party("mysubdomain.", firstMiddleware)
      mysubdomain.UseFunc(secondMiddleware)
      mysubdomain.Get("/", mysubdomainIndex)

      // per wildcard, dynamic subdomain
      dynamicSub := app.Party(".*", firstMiddleware, secondMiddleware)
      dynamicSub.Get("/", func(ctx *iris2.Context){
        ctx.Writef("Hello from subdomain: "+ ctx.Subdomain())
      })


    `iris2.ToHandler` converts(by wrapping) any `http.Handler/HandlerFunc` or
    `func(w http.ResponseWriter,r *http.Request, next http.HandlerFunc)` to an `iris2.HandlerFunc`.

    iris2.ToHandler(nativeNethttpHandler)

Let's convert the https://github.com/rs/cors net/http external middleware which returns a `next form` handler.


Example code:

      package main

      import (
      	"github.com/rs/cors"
      	"github.com/go-iris2/iris2"
      )

      // newCorsMiddleware returns a new cors middleware
      // with the provided options.
      func newCorsMiddleware() iris2.HandlerFunc {
      	options := cors.Options{
      		AllowedOrigins: []string{"*"},
      	}
      	handlerWithNext := cors.New(options).ServeHTTP

      	// this is the only func you will have to use if you're going
      	// to make use of any external net/http middleware.
      	// iris2.ToHandler converts the net/http middleware to an iris-compatible.
      	return iris2.ToHandler(handlerWithNext)
      }

      func main() {
      	app := iris2.New()

      	// Any registers a route to all http methods.
      	app.Any("/user", newCorsMiddleware(), func(ctx *iris2.Context) {
      		// ....
      	})

      	app.Listen(":8080")
      }



View Engine


Iris supports 5 template engines out-of-the-box, developers can still use any external golang template engine,
as `context.ResponseWriter` is an `io.Writer`.

All of these five template engines have common features with common API,
like Layout, Template Funcs, Party-specific layout, partial rendering and more.

      The standard html, based on github.com/go-iris2/iris2/template/tree/master/html
      its template parser is the golang.org/pkg/html/template/.

      Django, based ongithub.com/go-iris2/iris2/template/tree/master/django
      its template parser is the github.com/flosch/pongo2

      Pug(Jade), based on github.com/go-iris2/iris2/template/tree/master/pug
      its template parser is the github.com/Joker/jade

      Handlebars, based on github.com/go-iris2/iris2/template/tree/master/handlebars
      its template parser is the github.com/aymerick/raymond

      Amber, based on github.com/go-iris2/iris2/template/tree/master/amber
      its template parser is the github.com/eknkc/amber


Example code:

      package main

      import (
      	"github.com/go-iris2/iris2"
      	"github.com/go-iris2/iris2/adaptors/view" // <--- it contains all the template engines
      )

      func main() {
      	app := iris2.New(iris2.Configuration{Gzip: false, Charset: "UTF-8"}) // defaults to these

      	// - standard html  | view.HTML(...)
      	// - django         | view.Django(...)
      	// - pug(jade)      | view.Pug(...)
      	// - handlebars     | view.Handlebars(...)
      	// - amber          | view.Amber(...)
      	app.Adapt(view.HTML("./templates", ".html")) // <---- use the standard html

      	// default template funcs:
      	//
      	// - {{ url "mynamedroute" "pathParameter_ifneeded"} }
      	// - {{ urlpath "mynamedroute" "pathParameter_ifneeded" }}
      	// - {{ render "header.html" }}
      	// - {{ render_r "header.html" }} // partial relative path to current page
      	// - {{ yield }}
      	// - {{ current }}
      	//
      	// to adapt custom funcs, use:
      	app.Adapt(iris2.TemplateFuncsPolicy{"myfunc": func(s string) string {
      		return "hi " + s
      	}}) // usage inside template: {{ hi "kataras"}}

      	app.Get("/hi", func(ctx *iris2.Context) {
      		ctx.Render(
      			// the file name of the template relative to the './templates'.
      			"hi.html",
      			iris2.Map{"Name": "Iris"},
      			// the .Name inside the ./templates/hi.html,
      			// you can use any custom struct that you want to bind to the requested template.
      			iris2.Map{"gzip": false}, // set to true to enable gzip compression.
      		)

      	})

      	// http://127.0.0.1:8080/hi
      	app.Listen(":8080")
      }


View engine supports bundled(https://github.com/jteeuwen/go-bindata) template files too.
go-bindata gives you two functions, asset and assetNames,
these can be setted to each of the template engines using the `.Binary` func.

Example code:

      djangoEngine := view.Django("./templates", ".html")
      djangoEngine.Binary(asset, assetNames)
      app.Adapt(djangoEngine)

A real example can be found here: https://github.com/kataras/iris/tree/v6/adaptors/view/_examples/template_binary .

Enable auto-reloading of templates on each request. Useful while developers are in dev mode
as they no neeed to restart their app on every template edit.

Example code:


    pugEngine := view.Pug("./templates", ".jade")
    pugEngine.Reload(true) // <--- set to true to re-build the templates on each request.
    app.Adapt(pugEngine)


Each one of these template engines has different options located here: https://github.com/kataras/iris/tree/v6/adaptors/view .

That's the basics

But you should have a basic idea of the framework by now, we just scratched the surface.
If you enjoy what you just saw and want to learn more, please follow the below links:

* Examples: https://github.com/iris-contrib/examples

* Adaptors: https://github.com/go-iris2/iris2/tree/v6/adaptors

* Middleware: https://github.com/go-iris2/iris2/tree/v6/middleware and
* https://github.com/iris-contrib/middleware


*/
package iris2
